packages/bug-detectors/internal/prototype-pollution.ts

```diff
@@ -0,0 +1,565 @@
+								types.returnStatement(newVariable),
+							])
+						),
+						[]
+					);
+					instrumentationGuard.add(
+						TrackedExpressions.VARIABLE_DECLARATION,
+						path.node.init
+					);
+				}
+			},
+		},
+	};
+});
+
+// These objects will be used to detect prototype pollution.
+// Using global arrays for performance reasons.
+const BASIC_OBJECTS = [
+	{},
+	[],
+	"",
+	42,
+	true,
+	() => {
+		/**/
+	},
+];
+// The names are used in the Findings to print nicer messages.
+const BASIC_OBJECT_NAMES = [
+	"Object",
+	"Array",
+	"String",
+	"Number",
+	"Boolean",
+	"Function",
+];
+
+// Compute prototype snapshots of each selected basic object before any fuzz tests are run.
+// These snapshots are used to detect prototype pollution after each fuzz test.
+const BASIC_PROTO_SNAPSHOTS = computeBasicPrototypeSnapshots();
+
+function detectPrototypePollutionOfBasicObjects(): void {
+	const currentProtoSnapshots = computeBasicPrototypeSnapshots();
+	// Compare the current prototype snapshots of basic objects to the original ones.
+	for (let i = 0; i < BASIC_PROTO_SNAPSHOTS.length; i++) {
+		if (!currentProtoSnapshots[i]) {
+			reportFinding(
+				`Prototype Pollution: prototype of ${BASIC_OBJECT_NAMES[i]} changed.`
+			);
+			continue;
+		}
+		const equalityResult = protoSnapshotsEqual(
+			BASIC_PROTO_SNAPSHOTS[i],
+			currentProtoSnapshots[i]
+		);
+		if (!equalityResult.equal) {
+			reportFinding(
+				`Prototype Pollution: prototype of ${BASIC_OBJECT_NAMES[i]} changed.${equalityResult.message}`
+			);
+		}
+	}
+}
+
+registerAfterEachCallback(detectPrototypePollutionOfBasicObjects);
+
+// There are two main ways to pollute a prototype of an object:
+// 1. Changing a prototype's property using __proto__
+// 2. Changing a prototype's property using constructor.prototype
+// This dictionary adds these strings to the fuzzer dictionary.
+// Adding strings targeted at specific protocols (XML, HTTP, protobuf, etc.) will reduce the performance of the fuzzer,
+// because it will try strings from the wrong protocol. Therefore, it is advised to add protocol-specific strings
+// to the user dictionary for each fuzz test individually.
+addDictionary("__proto__", "constructor", "prototype", "constructor.prototype");
+
+type BasicProtoSnapshots = ProtoSnapshot[];
+
+function computeBasicPrototypeSnapshots(): BasicProtoSnapshots {
+	return BASIC_OBJECTS.map(getProtoSnapshot);
+}
+
+/**
+ * Checks if the object's proto contains any non-function properties. Function properties are ignored.
+ * @param obj The object to check.
+ * @param identifier The identifier of the object (used for printing a useful finding message).
+ * @param report Whether to report a finding if the object is a prototype pollution object.
+ */
+function detectPrototypePollution(
+	// eslint-disable-next-line  @typescript-eslint/no-explicit-any
+	obj: any,
+	identifier?: string,
+	report = true
+): boolean | void {
+	while (obj || obj === false || obj === 0 || obj === "") {
```

Comment:
This seems to exploit the truthiness behavior of certain types - could this be a more readable as `obj !== null && obj !== undefined && ...`?

Reply:
Correct, your version is better! Added.